utils/node-loader/src/batch_pool.rs

```diff
@@ -248,24 +257,55 @@ async fn process_events(
     let mut mailbox_added = BTreeSet::new();
 
     let results = loop {
-        let r = match api.events_since(block_hash, wait_for_events_blocks).await {
-            Ok(mut v) => {
-                // `gclient::EventProcessor` implementation on `IntoIterator` implementor clones `self` without mutating
-                // it, although `proc` and `proc_many` take mutable reference on `self`.
-                let mut mailbox_from_events = utils::capture_mailbox_messages(&api, &mut v)
-                    .await
-                    .expect("always valid by definition");
-                mailbox_added.append(&mut mailbox_from_events);
-
-                v.err_or_succeed_batch(messages.keys().copied()).await
+        let mut events = rx.recv().await?;
+        while events.block_hash() != block_hash {
+            tokio::time::sleep(Duration::new(0, 500)).await;
+            events = tokio::time::timeout(
+                Duration::from_millis(wait_for_events_millisec as u64),
+                rx.recv(),
+            )
+            .await
+            .map_err(|_| {
+                tracing::debug!("Timeout is reached while waiting for events");
+                anyhow!(utils::EVENTS_TIMEOUT_ERR_STR)
+            })??;
+        }
+
+        let r = {
+            let mut v = Vec::new();
+            let mut current_bh = events.block_hash();
+            let mut i = 0;
+            while i < wait_for_events_blocks {
+                if events.block_hash() != current_bh {
+                    current_bh = events.block_hash();
+                    i += 1;
+                }
+                for event in events.iter() {
+                    let event = event?.as_root_event::<gsdk::metadata::Event>()?;
+                    v.push(event);
+                }
+                tokio::time::sleep(Duration::new(0, 100)).await;
+                events = tokio::time::timeout(
+                    Duration::from_millis(wait_for_events_millisec as u64),
+                    rx.recv(),
+                )
+                .await
+                .map_err(|_| {
+                    tracing::debug!("Timeout is reached while waiting for events");
+                    anyhow!(utils::EVENTS_TIMEOUT_ERR_STR)
+                })??;
             }
-            Err(e) => Err(e),
-        };
 
-        if (gear_utils::now_millis() - now) as usize > wait_for_events_millisec {
-            tracing::debug!("Timeout is reached while waiting for events");
-            return Err(anyhow!(utils::EVENTS_TIMEOUT_ERR_STR));
-        }
+            let mut mailbox_from_events = utils::capture_mailbox_messages(&api, &mut v)
+                .await
+                .expect("always valid by definition");
+            mailbox_added.append(&mut mailbox_from_events);
+
+            tokio::task::block_in_place(|| {
+                tokio::runtime::Handle::current()
+                    .block_on(async { v.err_or_succeed_batch(messages.keys().copied()).await })
```

Comment:
btw, we can re-implement that on the loader, because we now have subxt, gclient, gsdk as dependencies here. What's more, `err_or_succeed_batch` doesn't have inside it any tasks to wait. It will return value immediately without yielding control to other thread. So it's worth having own implementation without any sync.

Reply:
DOne